fp/opt.go

```diff
@@ -0,0 +1,127 @@
+// Package opt implements the Option pattern from functional programming.
+// It provides a type-safe way to represent values that may or may not be present,
+// offering an alternative to nil pointers and helping to avoid nil pointer panics.
+package fp
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+// Opt represents a value that may or may not be present.
+// It serves as a container that explicitly handles the absence of a value.
+type Opt[T any] struct {
+	value *T
+}
+
+// Some creates an Opt containing a value.
+// Use this when you have a value that should be wrapped in an Opt.
+func Some[T any](value T) Opt[T] {
+	return Opt[T]{&value}
+}
+
+// None creates an empty Opt with no value.
+// Use this to represent the absence of a value.
+func None[T any]() Opt[T] {
+	return Opt[T]{}
+}
+
+// Filter returns the Opt with the value if the predicate returns true for the contained value,
+// otherwise returns None. If the Opt is already None, it returns None.
+func (o Opt[T]) Filter(predicate func(val T) bool) Opt[T] {
+	if o.value != nil && predicate(*o.value) {
+		return Some(*o.value)
+	}
+	return None[T]()
+}
+
+// Inspect executes the provided function on the value if it exists and returns the original Opt.
+// This is useful for side effects like logging without affecting the Opt chain.
+func (o Opt[T]) Inspect(f func(T) Opt[T]) Opt[T] {
```

Comment:
```suggestion
func (o Opt[T]) Inspect(f func(T)) Opt[T] {
```

We aren't consuming the result of `f()`, so I don't think it makes sense to have it return the type

Reply:
Add some examples on how Inspect would be useful with logging etc.

From the read it seems like a passive middleware function.